include/field.h

```diff
@@ -433,6 +435,26 @@ struct filter {
                                            const Store* store,
                                            const std::string& doc_id_prefix,
                                            std::vector<filter>& filters);
+
+    static Option<bool> parse_filter_query2(const std::string& filter_query,
+                                            const std::unordered_map<std::string, field>& search_schema,
+                                            const Store* store,
+                                            const std::string& doc_id_prefix,
+                                            filter_node_t* root);
+};
+
+struct filter_node_t {
+    std::string token;
```

Comment:
We can call this as `expression` since `token` means a word in the text in the rest of the project.